File: android/src/main/java/io/textile/rnmobile/TextileEvents.java
package io.textile.rnmobile;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import io.textile.pb.Model;
import io.textile.textile.FeedItemData;
import io.textile.textile.TextileEventListener;

class TextileEvents implements TextileEventListener {

    private DeviceEventManagerModule.RCTDeviceEventEmitter emitter;

    TextileEvents(final DeviceEventManagerModule.RCTDeviceEventEmitter emitter) {
        this.emitter = emitter;
    }

    @Override
    public void nodeStarted() {
        emitter.emit("NODE_STARTED", null);
    }

    @Override
    public void nodeFailedToStart(final Exception e) {
        emitter.emit("NODE_FAILED_TO_START", e.getMessage());
    }

    @Override
    public void nodeStopped() {
        emitter.emit("NODE_STOPPED", null);
    }

    @Override
    public void nodeFailedToStop(final Exception e) {
        emitter.emit("NODE_FAILED_TO_STOP", e.getMessage());
    }

    @Override
    public void nodeOnline() {
        emitter.emit("NODE_ONLINE", null);
    }

    @Override
    public void willStopNodeInBackgroundAfterDelay(final int seconds) {
        emitter.emit("WILL_STOP_NODE_IN_BACKGROUND_AFTER_DELAY", seconds);
    }

    @Override
    public void canceledPendingNodeStop() {
        emitter.emit("CANCELED_PENDING_NODE_STOP", null);
    }

    @Override
    public void notificationReceived(final Model.Notification notification) {
        emitter.emit("NOTIFICATION_RECEIVED", Util.encode(notification.toByteArray()));
    }

    @Override
    public void threadUpdateReceived(final String threadId, final FeedItemData feedItemData) {
        final WritableMap map = Arguments.createMap();
        map.putString("threadId", threadId);
        map.putString("block", feedItemData.block);
        map.putInt("type", feedItemData.type.ordinal());
        map.putString("data", Util.feedItemDataToBase64(feedItemData));
        emitter.emit("THREAD_UPDATE_RECEIVED", map);
    }

    @Override
    public void threadAdded(final String threadId) {
        emitter.emit("THREAD_ADDED", threadId);
    }

    @Override
    public void threadRemoved(final String threadId) {
        emitter.emit("THREAD_REMOVED", threadId);
    }

    @Override
    public void accountPeerAdded(final String peerId) {
        emitter.emit("ACCOUNT_PEER_ADDED", peerId);
    }

    @Override
    public void accountPeerRemoved(final String peerId) {
        emitter.emit("ACCOUNT_PEER_REMOVED", peerId);
    }

    @Override
    public void queryDone(final String queryId) {
        emitter.emit("QUERY_DONE", queryId);
    }

    @Override
    public void queryError(final String queryId, final Exception e) {
        final WritableMap map = new WritableNativeMap();
        map.putString("queryId", queryId);
        map.putString("error", e.getMessage());
        emitter.emit("QUERY_ERROR", map);
    }

    @Override
    public void clientThreadQueryResult(final String queryId, final Model.Thread thread) {
        final WritableMap map = new WritableNativeMap();
        map.putString("queryId", queryId);
        map.putString("data", Util.encode(thread.toByteArray()));
        emitter.emit("CLIENT_THREAD_QUERY_RESULT", map);
    }

    @Override
    public void contactQueryResult(final String queryId, final Model.Contact contact) {
        final WritableMap map = new WritableNativeMap();
        map.putString("queryId", queryId);
        map.putString("data", Util.encode(contact.toByteArray()));
        emitter.emit("CONTACT_QUERY_RESULT", map);
    }

    @Override
    public void syncUpdate(final Model.CafeSyncGroupStatus status) {
        emitter.emit("SYNC_UPDATE", Util.encode(status.toByteArray()));
    }

    @Override
    public void syncComplete(final Model.CafeSyncGroupStatus status) {
        emitter.emit("SYNC_COMPLETE", Util.encode(status.toByteArray()));
    }

    @Override
    public void syncFailed(final Model.CafeSyncGroupStatus status) {
        emitter.emit("SYNC_FAILED", Util.encode(status.toByteArray()));
    }
}

File: ios/TextileEvents.m
#import "TextileEvents.h"

#if __has_include(<React/RCTBridge.h>)
#import <React/RCTBridge.h>
#elif __has_include(“RCTBridge.h”)
#import “RCTBridge.h”
#else
#import “React/RCTBridge.h”
#endif

#import "TextileEvents.h"

@implementation TextileEvents

RCT_EXPORT_MODULE();

- (NSArray<NSString *> *)supportedEvents {
  return @[
           @"NODE_STARTED",
           @"NODE_FAILED_TO_START",
           @"NODE_STOPPED",
           @"NODE_FAILED_TO_STOP",
           @"NODE_ONLINE",
           @"WILL_STOP_NODE_IN_BACKGROUND_AFTER_DELAY",
           @"CANCELED_PENDING_NODE_STOP",
           @"NOTIFICATION_RECEIVED",
           @"THREAD_UPDATE_RECEIVED",
           @"THREAD_ADDED",
           @"THREAD_REMOVED",
           @"ACCOUNT_PEER_ADDED",
           @"ACCOUNT_PEER_REMOVED",
           @"QUERY_DONE",
           @"QUERY_ERROR",
           @"CLIENT_THREAD_QUERY_RESULT",
           @"CONTACT_QUERY_RESULT"
         ];
}

- (void)nodeStarted {
  [self sendEventWithName:@"NODE_STARTED" body:nil];
}

- (void)nodeFailedToStartWithError:(NSError *)error {
  [self sendEventWithName:@"NODE_FAILED_TO_START" body:error.description];
}

- (void)nodeStopped {
  [self sendEventWithName:@"NODE_STOPPED" body:nil];
}

- (void)nodeFailedToStopWithError:(NSError *)error {
  [self sendEventWithName:@"NODE_FAILED_TO_STOP" body:error.description];
}

- (void)nodeOnline {
  [self sendEventWithName:@"NODE_ONLINE" body:nil];
}

- (void)willStopNodeInBackgroundAfterDelay:(NSTimeInterval)seconds {
  [self sendEventWithName:@"WILL_STOP_NODE_IN_BACKGROUND_AFTER_DELAY" body:[NSNumber numberWithDouble:seconds]];
}

- (void)canceledPendingNodeStop {
  [self sendEventWithName:@"CANCELED_PENDING_NODE_STOP" body:nil];
}

- (void)notificationReceived:(Notification *)notification {
  [self sendEventWithName:@"NOTIFICATION_RECEIVED" body:[notification.data base64EncodedStringWithOptions:0]];
}

- (void)threadUpdateReceived:(FeedItem *)feedItem {
  [self sendEventWithName:@"THREAD_UPDATE_RECEIVED" body:[feedItem.data base64EncodedStringWithOptions:0]];
}

- (void)threadAdded:(NSString *)threadId {
  [self sendEventWithName:@"THREAD_ADDED" body:threadId];
}

- (void)threadRemoved:(NSString *)threadId {
  [self sendEventWithName:@"THREAD_REMOVED" body:threadId];
}

- (void)accountPeerAdded:(NSString *)peerId {
  [self sendEventWithName:@"ACCOUNT_PEER_ADDED" body:peerId];
}

- (void)accountPeerRemoved:(NSString *)peerId {
  [self sendEventWithName:@"ACCOUNT_PEER_REMOVED" body:peerId];
}

- (void)queryDone:(NSString *)queryId {
  [self sendEventWithName:@"QUERY_DONE" body:queryId];
}

- (void)queryError:(NSString *)queryId error:(NSError *)error {
  NSDictionary *body = @{
                         @"queryId": queryId,
                         @"error": error.description
                       };
  [self sendEventWithName:@"QUERY_ERROR" body:body];
}

- (void)clientThreadQueryResult:(NSString *)queryId thread:(Thread *)thread {
  NSDictionary *body = @{
                         @"queryId": queryId,
                         @"data": [thread.data base64EncodedStringWithOptions:0]
                       };
  [self sendEventWithName:@"CLIENT_THREAD_QUERY_RESULT" body:body];
}

- (void)contactQueryResult:(NSString *)queryId contact:(Contact *)contact {
  NSDictionary *body = @{
                         @"queryId": queryId,
                         @"data": [contact.data base64EncodedStringWithOptions:0]
                       };
  [self sendEventWithName:@"CONTACT_QUERY_RESULT" body:body];
}

@end

@implementation RCTBridge (TextileEvents)

- (TextileEvents *)textileEvents {
  return [self moduleForClass:[TextileEvents class]];
}

@end
